tests/unit/controllers/auth.test.ts

```diff
@@ -0,0 +1,179 @@
+            userUtil: userUtilStubs(),
+            jwt: jwtStubs()
+        };
+    });
+
+    beforeEach(() => {
+        mockRes.status = sinon.stub().returns(mockRes);
+        mockRes.json = sinon.stub().returns(mockRes);
+    });
+
+    afterEach(() => {
+        sinon.reset();
+    });
+
+    after(() => {
+        stubs.userDB.restore();
+        stubs.userValidator.restore();
+        stubs.userUtil.restore();
+        stubs.jwt.restore();
+    });
+
+    describe("Introspect", () => {
+
+        let req;
+        beforeEach(() => {
+            req = mockReq({
+                query: {
+                    token: testToken
+                }
+            });
+        });
+
+        it("status 200: returns a token with it's user", async () => {
+            stubs.userValidator.validationResult.returns({ isEmpty() { return true; } });
+            await authController.introspect(req, mockRes);
+            sinon.assert.calledWith(mockRes.status, statusCodes.SUCCESS);
+            sinon.assert.calledWith(mockRes.json, { active: true, user: { email: testUser.email, id: `${testUser._id}`, role: testUser.role, iat: testTokenIat } });
+        });
+
+        it("status 422: returns an appropriate response with validation errors", async () => {
+            const errorMsg = { status: statusCodes.MISSING_PARAMS, message: "query[token]: Missing 'token'" };
+            req.query.token = "not a token";
+            stubs.userValidator.validationResult.returns({
+                isEmpty() { return false; },
+                formatWith() {
+                    return {
+                        array() {
+                            return [errorMsg];
+                        }
+                    };
+                },
+            });
+            await authController.introspect(req, mockRes);
+            sinon.assert.calledOnce(stubs.userValidator.validationResult);
+            sinon.assert.calledWith(mockRes.status, statusCodes.MISSING_PARAMS);
+            sinon.assert.calledWith(mockRes.json, errorMsg);
+        });
+
+        it("status 401: returns unauthorized if invalid token is sent", async () => {
+            stubs.userValidator.validationResult.returns({ isEmpty() { return true; } });
+            req.query.token = "not a token";
+            await authController.introspect(req, mockRes);
+            sinon.assert.calledWith(mockRes.status, statusCodes.UNAUTHORIZED);
+        });
+    });
+
+    describe("Login", () => {
+
+        let req;
+        beforeEach(() => {
+            req = mockReq({
+                body: {
+                    email: testUser.email,
+                    password: "password"
+                }
+            });
+        });
+
+        it("status 200: returns new user", async () => {
+            stubs.userValidator.validationResult.returns({ isEmpty() { return true; } });
+            stubs.userDB.findByEmail.returns(testUser);
+            stubs.userUtil.codeforces.updateUserProblems.returns();
```

Comment:
This statement is redundant - for dev readability I think `.returns(null)` is preferable, or just omitting it since we should really only stub returns if we want it to return something IMO.